Dockerfile

```diff
@@ -1,10 +1,39 @@
-FROM golang
+# Build the spiffe-helper binary
+FROM --platform=$BUILDPLATFORM golang:1.21.5-alpine as base
+WORKDIR /workspace
 
-WORKDIR /build
-COPY . /build
-RUN CGO_ENABLED=0 go build -o spiffe-helper ./cmd/spiffe-helper
+# Cache deps before building and copying source so that we don't need to re-download as much
+# and so that source changes don't invalidate our downloaded layer
+COPY go.* ./
+RUN --mount=type=cache,target=/go/pkg/mod go mod download
+
+# Copy the go source
+COPY cmd/spiffe-helper/main.go cmd/spiffe-helper/main.go
+COPY pkg/ pkg/
+
+# xx is a helper for cross-compilation
+# when bumping to a new version analyze the new version for security issues
+# then use crane to lookup the digest of that version so we are immutable
+# crane digest tonistiigi/xx:1.3.0
+FROM --platform=${BUILDPLATFORM} tonistiigi/xx@sha256:904fe94f236d36d65aeb5a2462f88f2c537b8360475f6342e7599194f291fb7e AS xx
+
+FROM --platform=${BUILDPLATFORM} base as builder
+ARG TARGETPLATFORM
+ARG TARGETARCH
+
+ENV CGO_ENABLED=0
+COPY --link --from=xx / /
+RUN xx-go --wrap
+RUN --mount=type=cache,target=/root/.cache/go-build \
+    --mount=type=cache,target=/go/pkg/mod \
+    go build -o bin/spiffe-helper cmd/spiffe-helper/main.go
+
+# Use distroless as minimal base image to package the manager binary
+# Refer to https://github.com/GoogleContainerTools/distroless for more details
+#FROM gcr.io/distroless/static:nonroot
+FROM gcr.io/distroless/static AS spiffe-helper
+WORKDIR /
+COPY --link --from=builder /workspace/bin/spiffe-helper /spiffe-helper
 
-FROM scratch
-COPY --from=0 /build/spiffe-helper /spiffe-helper
 ENTRYPOINT ["/spiffe-helper"]
-CMD ["-config", "/etc/spiffe-helper.conf"]
```

Comment:
i'm not against this, but I'm worried about people using images generated by this, will start to fail without notification.
if we introduce this, we'll need to be loud in our changelog about this.
The good thing is that we are not publishing images, so we are safe

Reply:
We havent posted anything yet so it should be fine.

Reply:
Was thinking about this... If we go ahead and remove CMD, we can always add it back later to make it a little easier. So I don't think this should be blocking. Getting a release out, with an official container is much more important then this bit. We can go ahead and remove the CMD line and talk about it more in the future?